### ISSUE TITLE
Sequence length problem

### ISSUE THREAD

**Issue by zhangtianhong-1998** [huggingface/open-r1#579]
I found that I tried to run context 20k (32B) in the recipe with a single node and OOM got an error.
But in the context of success on 10240, I have some questions, because in my impression, under the conditions of data parallelism, multi-node and single-node on a single GPU should, occupancy does not seem to be different, is there a potential impact that causes the difference between multi-node and single-node on OOM

# Config for 16 nodes of 8 H100s with FSDP1
# Model arguments
model_name_or_path: /root/.cache/modelscope/hub/models/Qwen/Qwen2.5-32B-Instruct
model_revision: main
torch_dtype: bfloat16
attn_implementation: flash_attention_2

# Data training arguments
dataset_name: open-r1/codeforces-cots
dataset_config: solutions_decontaminated
dataset_num_proc: 12

# SFT trainer config
bf16: true
do_eval: false
eval_strategy: 'no'
gradient_accumulation_steps: 8
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
hub_always_push: true
hub_model_id: OlympicCoder-32B
hub_strategy: every_save
learning_rate: 4.0e-05
log_level: info
logging_steps: 1
logging_strategy: steps
lr_scheduler_type: cosine_with_min_lr
lr_scheduler_kwargs:
  min_lr_rate: 0.1
packing: false
max_grad_norm: 0.2
max_length: 22528  # we were unable to train at 32k due to OOM. See https://github.com/huggingface/transformers/issues/35983 for context parallelism support. 22528  10240
max_steps: -1
num_train_epochs: 10
optim: paged_adamw_8bit
output_dir: data/OlympicCoder-32B
overwrite_output_dir: true
per_device_eval_batch_size: 1
per_device_train_batch_size: 1
push_to_hub: false
report_to:
- wandb
save_only_model: true # needed to bypass FSDP errors with saving paged optimizers
save_strategy: epoch
save_total_limit: 1
seed: 42
use_liger: false # fails on multi-node
warmup_ratio: 0.03

**Comment by zhangtianhong-1998** [comment]
Using the following zero3.yaml configuration can enable the Qwen-2.5-32B model to be trained with a context length of 23552 on a single node with 8 H100 80GB GPUs.
------------------------------------- zero3.yaml-------------------------------------------
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  deepspeed_multinode_launcher: standard
  offload_optimizer_device: cpu
  # offload_optimizer_nvme_path: /media
  offload_param_device: cpu
  # offload_param_nvme_path: /media
  zero3_init_flag: true
  zero3_save_16bit_model: true
  zero_stage: 3

distributed_type: DEEPSPEED
downcast_bf16: 'no'
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 8
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: true
-------------------------------------recipes------------------------------------------
# Config for 16 nodes of 8 H100s with FSDP1
# Model arguments
model_name_or_path: /root/.cache/modelscope/hub/models/deepseek-ai/DeepSeek-R1-Distill-Qwen-32B
model_revision: main
torch_dtype: bfloat16
attn_implementation: flash_attention_2

# Data training arguments
dataset_name: open-r1/codeforces-cots
dataset_config: solutions_decontaminated
dataset_num_proc: 12

# SFT trainer config
bf16: true
do_eval: false
eval_strategy: 'no'
gradient_accumulation_steps: 16
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
hub_always_push: false
hub_model_id: DeepSeek-R1-Distill-Qwen-32B-sft-v2
hub_strategy: every_save
learning_rate: 1.0e-05
log_level: info
logging_steps: 2
logging_strategy: steps
lr_scheduler_type: cosine_with_min_lr
lr_scheduler_kwargs:
  min_lr_rate: 0.1
max_grad_norm: 0.2
# packing: true
max_length: 23552 # we were unable to train at 32k due to OOM. See https://github.com/huggingface/transformers/issues/35983 for context parallelism support. 
max_seq_length: 23552
max_steps: -1
num_train_epochs: 4
optim: paged_adamw_8bit
output_dir: /media/DeepSeek-R1-Distill-Qwen-32B-sft-v2
overwrite_output_dir: true
per_device_eval_batch_size: 1
per_device_train_batch_size: 1
push_to_hub: false
report_to:
- swanlab
wandb_project: DeepSeek-R1-Distill-Qwen-32B-sft-v2
save_only_model: true # needed to bypass FSDP errors with saving paged optimizers
save_strategy: epoch
save_total_limit: 4
seed: 42
use_liger: false # fails on multi-node
warmup_ratio: 0.02

**Comment by zhangtianhong-1998** [comment]
I previously didn't have a deep understanding of Zero3 and FSDP. After relearning and modifying the parameters, I've made some progress.

**Comment by deeplearning-LI** [comment]
Hello, I am facing the same problem. I use one node and 8 GPUs to train 32B, but max_seq_length only supports less than 1k. Has your problem been solved?

**Comment by zhangtianhong-1998** [comment]
> Hello, I am facing the same problem. I use one node and 8 GPUs to train 32B, but max_seq_length only supports less than 1k. Has your problem been solved?
Using Zero3 with offloading weights to the CPU, based on the parameters provided above, and with the max_seq_length reduced to 22000 for Qwen3。

**Comment by zhangtianhong-1998** [comment]
> I have already offloaded the parameters to the CPU, and the maximum sequence length is 8192 and it doesn't work.
> 
> _deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: none offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3_
> 
> Is there any other solution?

debug: false
deepspeed_config:
  deepspeed_multinode_launcher: standard
  offload_optimizer_device: cpu
  offload_param_device: cpu
  zero3_init_flag: true
  zero3_save_16bit_model: true
  zero_stage: 3

distributed_type: DEEPSPEED
downcast_bf16: 'no'
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 8
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: true

**Comment by zhangtianhong-1998** [comment]
> > > I have already offloaded the parameters to the CPU, and the maximum sequence length is 8192 and it doesn't work.
> > > _deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: none offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3_
> > > Is there any other solution?
> > 
> > 
> > debug: false deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: cpu offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3
> > distributed_type: DEEPSPEED downcast_bf16: 'no' machine_rank: 0 main_training_function: main mixed_precision: bf16 num_machines: 1 num_processes: 8 rdzv_backend: static same_network: true tpu_env: [] tpu_use_cluster: false tpu_use_sudo: false use_cpu: true
> 
> Based on these settings, how much can your max_seq_length be set to? Are you using peft? According to your parameter settings, my eight A800s still have OOM.

max_seq_length  = 23552
every Gpu Must be 80GB

**Comment by deeplearning-LI** [comment]
> > > > I have already offloaded the parameters to the CPU, and the maximum sequence length is 8192 and it doesn't work.
> > > > _deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: none offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3_
> > > > Is there any other solution?
> > > 
> > > 
> > > debug: false deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: cpu offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3
> > > distributed_type: DEEPSPEED downcast_bf16: 'no' machine_rank: 0 main_training_function: main mixed_precision: bf16 num_machines: 1 num_processes: 8 rdzv_backend: static same_network: true tpu_env: [] tpu_use_cluster: false tpu_use_sudo: false use_cpu: true
> > 
> > 
> > Based on these settings, how much can your max_seq_length be set to? Are you using peft? According to your parameter settings, my eight A800s still have OOM.
> 
> max_seq_length = 23552 every Gpu Must be 80GB

Hi, Is your data so long? Or do you only calculate the loss of part of the sequence?

**Comment by zhangtianhong-1998** [comment]
> > > > > I have already offloaded the parameters to the CPU, and the maximum sequence length is 8192 and it doesn't work.
> > > > > _deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: none offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3_
> > > > > Is there any other solution?
> > > > 
> > > > 
> > > > debug: false deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: cpu offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3
> > > > distributed_type: DEEPSPEED downcast_bf16: 'no' machine_rank: 0 main_training_function: main mixed_precision: bf16 num_machines: 1 num_processes: 8 rdzv_backend: static same_network: true tpu_env: [] tpu_use_cluster: false tpu_use_sudo: false use_cpu: true
> > > 
> > > 
> > > Based on these settings, how much can your max_seq_length be set to? Are you using peft? According to your parameter settings, my eight A800s still have OOM.
> > 
> > 
> > max_seq_length = 23552 every Gpu Must be 80GB
> 
> Hi, Is your data so long? Or do you only calculate the loss of part of the sequence?

Yes, it will be this long after including the data of the thought chain. Besides, by default, the trl framework is used here. I think internally, only the response loss should be calculated

**Comment by deeplearning-LI** [comment]
> > > > > > I have already offloaded the parameters to the CPU, and the maximum sequence length is 8192 and it doesn't work.
> > > > > > _deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: none offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3_
> > > > > > Is there any other solution?
> > > > > 
> > > > > 
> > > > > debug: false deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: cpu offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3
> > > > > distributed_type: DEEPSPEED downcast_bf16: 'no' machine_rank: 0 main_training_function: main mixed_precision: bf16 num_machines: 1 num_processes: 8 rdzv_backend: static same_network: true tpu_env: [] tpu_use_cluster: false tpu_use_sudo: false use_cpu: true
> > > > 
> > > > 
> > > > Based on these settings, how much can your max_seq_length be set to? Are you using peft? According to your parameter settings, my eight A800s still have OOM.
> > > 
> > > 
> > > max_seq_length = 23552 every Gpu Must be 80GB
> > 
> > 
> > Hi, Is your data so long? Or do you only calculate the loss of part of the sequence?
> 
> Yes, it will be this long after including the data of the thought chain. Besides, by default, the trl framework is used here. I think internally, only the response loss should be calculated

Thanks for your reply.

**Comment by deeplearning-LI** [comment]
> > > > > > > I have already offloaded the parameters to the CPU, and the maximum sequence length is 8192 and it doesn't work.
> > > > > > > _deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: none offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3_
> > > > > > > Is there any other solution?
> > > > > > 
> > > > > > 
> > > > > > debug: false deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: cpu offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3
> > > > > > distributed_type: DEEPSPEED downcast_bf16: 'no' machine_rank: 0 main_training_function: main mixed_precision: bf16 num_machines: 1 num_processes: 8 rdzv_backend: static same_network: true tpu_env: [] tpu_use_cluster: false tpu_use_sudo: false use_cpu: true
> > > > > 
> > > > > 
> > > > > Based on these settings, how much can your max_seq_length be set to? Are you using peft? According to your parameter settings, my eight A800s still have OOM.
> > > > 
> > > > 
> > > > max_seq_length = 23552 every Gpu Must be 80GB
> > > 
> > > 
> > > Hi, Is your data so long? Or do you only calculate the loss of part of the sequence?
> > 
> > 
> > Yes, it will be this long after including the data of the thought chain. Besides, by default, the trl framework is used here. I think internally, only the response loss should be calculated
> 
> Thanks for your reply.

I rethought it, if you calculate the loss of cot, 1*8 gpus should not be enough, the length of tokens involved in the loss may exceed 10k. If you can really train it and don't use PEFT, I think there is a bug.

**Comment by zhangtianhong-1998** [comment]
> > > > > > > > I have already offloaded the parameters to the CPU, and the maximum sequence length is 8192 and it doesn't work.
> > > > > > > > _deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: none offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3_
> > > > > > > > Is there any other solution?
> > > > > > > 
> > > > > > > 
> > > > > > > debug: false deepspeed_config: deepspeed_multinode_launcher: standard offload_optimizer_device: cpu offload_param_device: cpu zero3_init_flag: true zero3_save_16bit_model: true zero_stage: 3
> > > > > > > distributed_type: DEEPSPEED downcast_bf16: 'no' machine_rank: 0 main_training_function: main mixed_precision: bf16 num_machines: 1 num_processes: 8 rdzv_backend: static same_network: true tpu_env: [] tpu_use_cluster: false tpu_use_sudo: false use_cpu: true
> > > > > > 
> > > > > > 
> > > > > > Based on these settings, how much can your max_seq_length be set to? Are you using peft? According to your parameter settings, my eight A800s still have OOM.
> > > > > 
> > > > > 
> > > > > max_seq_length = 23552 every Gpu Must be 80GB
> > > > 
> > > > 
> > > > Hi, Is your data so long? Or do you only calculate the loss of part of the sequence?
> > > 
> > > 
> > > Yes, it will be this long after including the data of the thought chain. Besides, by default, the trl framework is used here. I think internally, only the response loss should be calculated
> > 
> > 
> > Thanks for your reply.
> 
> I rethought it, if you calculate the loss of cot, 1*8 gpus should not be enough, the length of tokens involved in the loss may exceed 10k. If you can really train it and don't use PEFT, I think there is a bug.

The specific implementation needs to start from the source code of TRL. Due to encapsulation reasons, I'm not sure either. I only implement it as recommended by open-r1
